### PR TITLE
[FLINK-28459][table-planner] Supports non-atomic CREATE TABLE AS SELECT

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

* Supports non-atomic CREATE TABLE AS SELECT

## Brief change log

Modify the SqlCreateTableConverter and SqlToOperationConverter to support non-atomic CREATE TABLE AS SELECT

## Verifying this change

add ITCase
* org.apache.flink.table.planner.runtime.batch.sql.TableSinkITCase#testCreateTableAsSelect
* org.apache.flink.table.planner.runtime.stream.sql.TableSinkITCase#testCreateTableAsSelect

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by Tartarus0zm** [comment]
@flinkbot run azure

**Review by lsyldliu — COMMENTED** [review]
@Tartarus0zm Thanks for your contribution, I left some comments.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSinkITCase.scala [hunk 1]
```diff
@@ -192,4 +193,30 @@ class TableSinkITCase(mode: StateBackendMode) extends StreamingWithStateTestBase
...
+                    |""".stripMargin)
+      .await()
+    val actual = TestValuesTableFactory.getResults("MyCtasTable")
+    val expected = List(
+      "+I[jason, 1]",
+      "+I[jason, 1]",
+      "+I[jason, 1]",
+      "+I[jason, 1]"
+    )
+    Assertions.assertThat(actual.sorted).isEqualTo(expected.sorted)
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSinkITCase.scala:220]: [flink code style](https://flink.apache.org/contributing/code-style-and-quality-common.html#avoid-mockito---use-reusable-test-implementations)
Refer to this coding specification

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TableSinkITCase.scala [hunk 2]
```diff
@@ -94,4 +94,36 @@ class TableSinkITCase extends BatchTestBase {
...
+                     |  'connector' = 'filesystem',
+                     |  'format' = 'testcsv',
+                     |  'path' = '$resultPath'
+                     |) AS
+                     | SELECT * FROM MyTable
+       """.stripMargin)
+      .await()
+    val expected = Seq("1,1,Hi", "2,2,Hello", "3,2,Hello world")
+    val result = TableTestUtil.readFromFile(resultPath)
+    Assertions.assertThat(result.sorted).isEqualTo(expected.sorted)
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TableSinkITCase.scala:127]: ditto

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TableSinkITCase.scala [hunk 3]
```diff
@@ -94,4 +94,36 @@ class TableSinkITCase extends BatchTestBase {
     tEnv.getConfig.set(CollectSinkOperatorFactory.MAX_BATCH_SIZE, MemorySize.parse("1kb"))
     checkResult("SELECT 1", Seq(row(1)))
   }
+
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TableSinkITCase.scala:97]: explain does not support CreateTableASOperation yet.
[FLINK-28770](https://issues.apache.org/jira/browse/FLINK-28770) will follow up to resolve

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TableSinkITCase.scala [hunk 4]
```diff
@@ -94,4 +94,36 @@ class TableSinkITCase extends BatchTestBase {
...
+  @Test
+  def testCreateTableAsSelect(): Unit = {
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/TableSinkITCase.scala:99]: Add TODO in UT testCreateTableAsSelectWithoutOptions

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSinkITCase.scala [hunk 5]
```diff
@@ -192,4 +193,41 @@ class TableSinkITCase(mode: StateBackendMode) extends StreamingWithStateTestBase
...
+  }
+
+  @Test
+  def testCreateTableAsSelectWithoutOptions(): Unit = {
+    // TODO CTAS supports ManagedTable
+    Assertions
```
Comment (lsyldliu) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSinkITCase.scala:226]: Please also explain the reason?

**Review by lsyldliu — COMMENTED** [review]
Please also explain why not can add ctas plan related test in comment.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSinkITCase.scala (hunk 5 above)
Comment (Tartarus0zm) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSinkITCase.scala:226]: If the connector option is not specified, Flink will creates a Managed table.
Managed table relies on checkpoint to commit and the data is visible only after commit.
Managed table requires two layers of log storage and file storage and depends on the flink table store, CTAS will support Managed Table in the future.

**Review by lsyldliu — APPROVED** [review]

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**